posts/nix-vs-docker-i.md

```diff
@@ -0,0 +1,260 @@
+---
+title: "Nix vs Docker: Containers Without the Bloat"
+description: "Introduction to the NixOS testing framework with flakes"
+date: 2025-04-20
+keywords: ["nix", "nixos", "docker"]
+---
+
+# Nix vs Docker: Containers Without the Bloat
+
+Docker brought containers into the mainstream by simplifying the idea of
+packaging an application and all its dependencies into a single unit. You write
+a Dockerfile, build an image, and then you ship it. The abstraction is clean at
+first glance, but under the hood, Docker makes some serious compromises. It
+depends on mutable state, relies on an imperative build model, and often
+produces inconsistent or bloated results.
+
+Nix, by contrast, approaches the same problem from a fundamentally different
+angle. It's not a container engine. It doesn't emulate layers. Instead, it
+treats builds like pure functions. Given a set of inputs, Nix evaluates a
+dependency graph and produces an immutable output. And it does this
+reproducibly, deterministically, and without requiring a privileged daemon to
+run in the background.
+
+Both tools aim to isolate environments and improve portability, but the
+underlying models are so different that it's not a fair one-to-one comparison.
+Still, for a growing number of use cases, Nix doesn't just replace Docker. It
+does a better job. Nix, after all, is a superset of Docker.
+
+## The Build Model
```

Comment:
Could mention how Docker layers are a linked list of dependencies, whereas Nix models dependencies as a DAG, which is what they actually are. Dockerfiles force you to apply a toposort to your dep graph.